### PR TITLE
feat(shwap): Add ODS file

### PR DESCRIPTION
**Pull request by walldiss**
This PR adds ods file, that implements eds.Accessor interface. The file stores ods part of eds on disk and lazily reads data upon request. If requested data is from Q4, it reads full ods in single read and stores it in-memory for later re-use.

Based on https://github.com/celestiaorg/celestia-node/pull/3425

### CODE REVIEW

**Comment by walldiss** [comment]
Lint will be get fixed after linter upgrade https://github.com/celestiaorg/celestia-node/pull/3483

File: store/file/codec_test.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/klauspost/reedsolomon"
+	"github.com/stretchr/testify/require"
+
+	"github.com/celestiaorg/celestia-node/share/sharetest"
+)
+
+func BenchmarkCodec(b *testing.B) {
+	minSize, maxSize := 32, 128
+
+	for size := minSize; size <= maxSize; size *= 2 {
```
Comment (cristaloleg) [inline store/file/codec_test.go:16]: IMO feels like this should be inside every (sub-) benchmark to make them isolated and more visible relation to `size` variable.

Comment (cristaloleg) [reply store/file/codec_test.go:16]: Ah...we use it in benchmark name...so (probably) ignore my comment.

File: store/file/header.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		squareSize: binary.LittleEndian.Uint16(bytesHeader[4:6]),
+		datahash:   make([]byte, 32),
+	}
+
+	copy(h.datahash, bytesHeader[6:6+32])
+	return h, err
+}
+
+func writeLenEncoded(w io.Writer, data []byte) (int64, error) {
+	_, err := w.Write([]byte{byte(len(data))})
```
Comment (cristaloleg) [inline store/file/header.go:71]: Waaaait, are we writing just the lower 8bits of `len(data)` here?

File: store/file/square.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	axisType rsmt2d.Axis,
+	axisIdx int,
+) ([]share.Share, error) {
+	shares := make([]share.Share, s.size())
+
+	// extend opposite half of the square while collecting Shares for the first half of required axis
+	g, ctx := errgroup.WithContext(ctx)
+	opposite := oppositeAxis(axisType)
+	for i := 0; i < s.size(); i++ {
+		i := i
```
Comment (cristaloleg) [inline store/file/square.go:95]: Starting with Go 1.22 we don't need that https://go.dev/blog/loopvar-preview

File: store/file/ods_test.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+// BenchmarkShareFromOdsFile/Size:128/Axis:col/squareHalf:first(original)-10        	    2114	    514642 ns/op
+// BenchmarkShareFromOdsFile/Size:128/Axis:col/squareHalf:second(extended)-10       	     373	   3068104 ns/op
+func BenchmarkShareFromOdsFile(b *testing.B) {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	b.Cleanup(cancel)
+
+	minSize, maxSize := 32, 128
+	dir := b.TempDir()
+
+	newFile := func(size int) eds.Accessor {
```
Comment (cristaloleg) [inline store/file/ods_test.go:137]: Looks like it's the same as above. Do we need to keep this as a param or just leave on eds testing funcs?

File: store/file/header.go (hunk 2 above)
Comment (cristaloleg) [reply store/file/header.go:71]: Ok, looks like we always(?) write 64 bytes (`headerSize`), am I right?

File: store/file/header.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	if err != nil {
+		return 0, err
+	}
+	return io.Copy(w, bytes.NewBuffer(data))
```
Comment (cristaloleg) [inline store/file/header.go:75]: `w.Write(data)` ?

File: store/file/ods.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("writing ODS file: %w", err)
+	}
+
+	// TODO: fill ods field with data from eds
+	return &OdsFile{
+		path: path,
+		fl:   f,
+		hdr:  h,
+	}, f.Sync()
```
Comment (cristaloleg) [inline store/file/ods.go:76]: I read this file 3 times and only on the last iteration I noticed this `f.Sync`, what about to move it right after `if` above?

File: store/file/header.go (hunk 2 above)
Comment (walldiss) [reply store/file/header.go:71]: It writes len of data into first byte. Right now we read 64 bytes always no matter what. In future header struct might grow and new fields might be added. It would require old storages to be upgraded to bigger header or lenEncoded version of it. So lenEncoding is added to prevent future break of full file. Otherwise we would need to rewrite all files fully on file.header upgrade. 
TLDR: It doesn't do much as of now, but will be needed if header struct grow to keep backwards compatibility.

File: store/file/header.go (hunk 5 above)
Comment (walldiss) [reply store/file/header.go:75]: I don't want to manage amount bytes written manually and prefer io.Copy to do the job

Comment (cristaloleg) [reply store/file/header.go:75]: In my example you don't manage how many bytes you're writing. If it will fail - it will.

File: store/file/header.go (hunk 2 above)
Comment (cristaloleg) [reply store/file/header.go:71]: Sounds like a good place for a) write `int32` which will be enough for any sane version (also, it's header, I bet it will be 1Gb) or b) specify header version with byte counter.

Anyway, both of such solutions will make simpler our migration in the future (even if it will happen).

File: store/file/header.go (hunk 5 above)
Comment (walldiss) [reply store/file/header.go:75]: For example, if you write 100 bytes with w.Write(data), it can write first 10 bytes out of 100 and return `10, nil`. While it is not an error, you need yo manually write the rest to the writer. I don't want to do it manually.

Comment (cristaloleg) [reply store/file/header.go:75]: As per `io.Writer` documentation (which sadly doesn't give any guarantees on the interface implementation, oof) it should work as: `It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early.`

So, if it writer less than `len(p)` we should have an error otherwise everything should be written fully.

I don't see a good argument to continue this thread now, let's leave it as `io.Copy + bytes.Buffer`.

Comment (quasilyte) [reply store/file/header.go:75]: If one doesn't trust `Writer` inteface to return a non-nil error when n is less than `len(data)`, why one trusts `WriterTo` interface to be any better?

`io.Copy` will recognize that `bytes.Buffer` implements `io.WriterTo` method and forward the call directly:
https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/io/io.go;l=411
https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/bytes/buffer.go;l=258

Here is the WriterTo documentation:
https://pkg.go.dev/io#WriterTo
It doesn't give any contracts while io.Writer docs require this:
> Write must return a non-nil error if it returns n < len(p). Write must not modify the slice data, even temporarily. 

I would at least give it a second thought. The code right now is more fragile (it relies on something that is not guaranteed while avoiding something with a specified contract) plus it does more job (creates a new bytes buffer, for example) than necessary.

Comment (walldiss) [reply store/file/header.go:75]: `WriteTo` implementation of bytes.Buffer does give guarantee of full write, and in case when not full data has been written it will return io.ErrShortWrite. Nonetheless I agree, that in discussed snippet ordinary `Write` should be sufficient and will be simplier. I'll will update code accordingly

File: store/file/ods_test.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
+//nolint:goconst
```
Comment (walldiss) [inline store/file/ods_test.go:1]: This will be ineffective after upgrade to 1.59 golangci-lint and will be removed

File: store/file/header.go (hunk 2 above)
Comment (walldiss) [reply store/file/header.go]: I've removed length encoding in favour of headerVersion

**Review by Wondertan — COMMENTED** [review]
First look, still have to go over the logic in square and the file

File: share/shwap/row_namespace_data.go [hunk 8]
```diff
@@ -164,10 +164,14 @@ func (rnd RowNamespaceData) Validate(dah *share.Root, namespace share.Namespace,
...
 	leaves := make([][]byte, 0, len(rnd.Shares))
-	for _, shr := range rnd.Shares {
-		namespaceBytes := share.GetNamespace(shr)
-		leaves = append(leaves, append(namespaceBytes, shr...))
+	for _, sh := range rnd.Shares {
+		namespaceBytes := share.GetNamespace(sh)
+		leave := make([]byte, len(sh)+len(namespaceBytes))
+		copy(leave, namespaceBytes)
+		copy(leave[len(namespaceBytes):], sh)
+		leaves = append(leaves, leave)
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/shwap/row_namespace_data.go:172]: I like that you noticed. There was a very nasty mutations bug that took me few hours to fix.

File: store/file/codec.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
+package file
+
+import (
+	"sync"
+
+	"github.com/klauspost/reedsolomon"
+)
+
+var codec Codec
```
[the remark this answers is not in the document]
Comment (walldiss) [reply store/file/codec.go:9]: Access to ReconstructSome

File: store/file/header.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+type headerVersion uint8
+
+const (
+	headerVersionV0 headerVersion = 1
+	headerVOSize    int           = 40
+)
+
+type headerV0 struct {
```
[the remark this answers is not in the document]
Comment (walldiss) [reply store/file/header.go:18]: tye field is fileType, not a header version

File: store/file/square.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+	square := make(square, odsLn)
+	for i := range square {
+		square[i] = make([]share.Share, odsLn)
+		for j := range square[i] {
+			square[i][j] = make(share.Share, shareSize)
+		}
+	}
+
+	// TODO(@walldiss): run benchmark to find optimal size for this buffer
```
Comment (Wondertan) [inline store/file/square.go:32]: I would park this in the issue, so that some rando interested in low-level things could contribute

Comment (Wondertan) [reply store/file/square.go:32]: It might not be an issue to be created now, but we could make a little list of low-prior issues to be posted to GH. This one for example and the Size returning error

File: store/file/square.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+	"io"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+)
+
+type square [][]share.Share
```
Comment (Wondertan) [inline store/file/square.go:16]: Is this PR used only for ODS? Should we call it so, then? Or will it be used for EDS as well?

Comment (Wondertan) [reply store/file/square.go:16]: Or at least the `readSquare` func because its hardcoded to read the ODS, but square mean both ODS and EDS. Need to clarify

File: store/file/square.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+func readSquare(r io.Reader, shareSize, edsSize int) (square, error) {
+	odsLn := edsSize / 2
+
+	square := make(square, odsLn)
+	for i := range square {
+		square[i] = make([]share.Share, odsLn)
+		for j := range square[i] {
+			square[i][j] = make(share.Share, shareSize)
+		}
+	}
```
Comment (Wondertan) [inline store/file/square.go:30]: Don't we want to read the square by Row number of reads instead of Row*Col reads? Not sure why you preferred reading share-by-share instead of row-by-row

File: store/file/square.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+func (s square) shares() ([]share.Share, error) {
+	shares := make([]share.Share, 0, s.size()*s.size())
+	for _, row := range s {
+		shares = append(shares, row...)
+	}
+	return shares, nil
+}
+
+func (s square) computeAxisHalf(
```
Comment (Wondertan) [inline store/file/square.go:84]: Isn't this a full axis?

Comment (Wondertan) [reply store/file/square.go:84]: Ohh, compute the opposite half, basically parity, because the square stores the ODS. This needs to be clarified in the comment or by a better function name, I believe.

File: store/file/square.go [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+			target := make([]bool, s.size()*2)
+			target[axisIdx] = true
+
+			err = enc.ReconstructSome(shards, target)
+			if err != nil {
+				return fmt.Errorf("reconstruct some: %w", err)
+			}
+
+			shares[i] = shards[axisIdx]
```
Comment (Wondertan) [inline store/file/square.go:117]: Shouldn't we cache this on the square? Is this what the cache file is for?

File: store/file/square.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+				return err
+			}
+
+			enc, err := codec.Encoder(s.size() * 2)
+			if err != nil {
+				return fmt.Errorf("encoder: %w", err)
+			}
+
+			shards := make([][]byte, s.size()*2)
+			copy(shards, original)
```
Comment (Wondertan) [inline store/file/square.go:107]: Why is copy necessary?

File: store/file/square.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+			target := make([]bool, s.size()*2)
+			target[axisIdx] = true
+
+			err = enc.ReconstructSome(shards, target)
+			if err != nil {
+				return fmt.Errorf("reconstruct some: %w", err)
+			}
```
Comment (Wondertan) [inline store/file/square.go:115]: This actually doesn't need a reconstructSome, because it always does the Reconstruct operation as it recovers the full half

Comment (Wondertan) [reply store/file/square.go:115]: So can we avoid spending time on Some?

File: store/file/square.go [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+	for i := 0; i < s.size(); i++ {
+		g.Go(func() error {
+			original, err := s.axisHalf(ctx, opposite, i)
+			if err != nil {
+				return err
+			}
+
+			enc, err := codec.Encoder(s.size() * 2)
+			if err != nil {
+				return fmt.Errorf("encoder: %w", err)
```
Comment (Wondertan) [inline store/file/square.go:103]:
```suggestion
				return fmt.Errorf("getting encoder: %w", err)
```

File: store/file/ods.go [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+		return eds.AxisHalf{}, err
+	}
+
+	shares, err := f.ods.computeAxisHalf(ctx, axisType, axisIdx)
+	if err != nil {
+		return eds.AxisHalf{}, fmt.Errorf("computing axis half: %w", err)
+	}
+	return eds.AxisHalf{
+		Shares:   shares,
+		IsParity: false,
```
Comment (Wondertan) [inline store/file/ods.go:170]: This should be true

File: store/file/ods.go [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+	if ods != nil {
+		return f.ods.axisHalf(context.Background(), axisType, axisIdx)
+	}
+
+	switch axisType {
+	case rsmt2d.Col:
+		return f.readCol(axisIdx, 0)
+	case rsmt2d.Row:
+		return f.readRow(axisIdx)
+	}
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply store/file/ods.go:209]: Ah, I guess this is gonna be responsibility of the cache file

File: store/file/ods.go [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+
+	shares, err := f.ods.computeAxisHalf(ctx, axisType, axisIdx)
+	if err != nil {
+		return eds.AxisHalf{}, fmt.Errorf("computing axis half: %w", err)
+	}
+	return eds.AxisHalf{
+		Shares:   shares,
+		IsParity: false,
+	}, nil
```
Comment (Wondertan) [inline store/file/ods.go:171]: I am a bit concerned that this gonna recompute halfs twice:
* computeAxisHalf: square -> to parity half
* AxisHalf: parity half -> original
  * When original is already in the square

File: store/file/ods.go [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+		hdr:  h,
+		fl:   f,
+	}, nil
+}
+
+// CreateOdsFile creates a new file. File has to be closed after usage.
+func CreateOdsFile(
+	path string,
+	datahash share.DataHash,
+	eds *rsmt2d.ExtendedDataSquare,
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply store/file/ods.go:58]: In sync: we don't need that right now, but we may return to this later if needed, e.g. storage operating over Accessors isntead of the concretion

File: store/file/ods.go [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+	}
+
+	return shwap.SampleFromShares(axis, axisType, axisIdx, shrIdx)
+}
+
+// AxisHalf returns half of shares axis of the given type and index. Side is determined by
+// implementation. Implementations should indicate the side in the returned AxisHalf.
+func (f *OdsFile) AxisHalf(ctx context.Context, axisType rsmt2d.Axis, axisIdx int) (eds.AxisHalf, error) {
+	// read axis from file if axisis row and from top half of the square or if axis is column and from
+	// left half of the square
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply store/file/ods.go:146]: nvm there is column case in the comment later as well

File: store/file/square.go (hunk 12 above)
Comment (Wondertan) [reply store/file/square.go:16]: Sync: call it a `quadrant`

File: store/file/square.go (hunk 14 above)
Comment (Wondertan) [reply store/file/square.go:84]: sync: comment with a picture

Comment (Wondertan) [reply store/file/square.go:84]: and generally more comments

File: store/file/square.go (hunk 15 above)
Comment (Wondertan) [reply store/file/square.go:117]: Yes

File: store/file/square.go (hunk 16 above)
Comment (Wondertan) [reply store/file/square.go:107]: Because we need to allocate the full row here for codec's API?

Comment (Wondertan) [reply store/file/square.go:107]: Needs more comments though

File: store/file/square.go (hunk 17 above)
Comment (Wondertan) [reply store/file/square.go:115]: My comment was wrong. Some is needed to recompute particular opposite shares, so we better off keeping it

File: store/file/ods.go (hunk 19 above)
Comment (Wondertan) [reply store/file/ods.go:170]: That is not true. I got confused because objectively, this is parity, but relative to the axis in Q2, this is original data.

File: store/file/ods.go (hunk 21 above)
Comment (Wondertan) [reply store/file/ods.go:171]: This comment is also wrong and based on a misunderstanding of computeAxisHalf works

File: store/file/square.go [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+	shares := make([]share.Share, 0, s.size()*s.size())
+	for _, row := range s {
+		shares = append(shares, row...)
+	}
+	return shares, nil
+}
+
+// TODO(@walldiss): make comments with diagram of computed axis. Add more comment on actual algo
+func (s square) computeAxisHalf(
+	ctx context.Context,
```
Comment (Wondertan) [inline store/file/square.go:86]: We don't actually need context here

File: store/file/square.go [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+		}
+	}
+	return square, nil
+}
+
+func (s square) size() int {
+	return len(s)
+}
+
+func (s square) axisHalf(_ context.Context, axisType rsmt2d.Axis, axisIdx int) ([]share.Share, error) {
```
Comment (Wondertan) [inline store/file/square.go:54]: Use AxisHalf type instead

File: store/file/square.go [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+	// construct half column from row ordered square
+	col := make([]share.Share, s.size())
+	for i := 0; i < s.size(); i++ {
+		col[i] = s[i][axisIdx]
+	}
+	return col, nil
+}
+
+func (s square) shares() ([]share.Share, error) {
```
Comment (Wondertan) [inline store/file/square.go:76]: Put above axisHalf

File: store/file/ods.go [hunk 27]
```diff
@@ -193,21 +187,29 @@ func (f *ODSFile) Shares(context.Context) ([]share.Share, error) {
...
 	case rsmt2d.Row:
-		return f.readRow(axisIdx)
+		row, err := f.readRow(axisIdx)
+		return eds.AxisHalf{
+			Shares:   row,
+			IsParity: false,
+		}, err
 	}
-	return nil, fmt.Errorf("unknown axis")
+	return eds.AxisHalf{}, fmt.Errorf("unknown axis")
```
Comment (Wondertan) [inline store/file/ods.go]: uber optional nit: dedup

File: store/file/ods.go [hunk 28]
```diff
@@ -233,7 +235,7 @@ func (f *ODSFile) readODS() error {
 
 func (f *ODSFile) readRow(idx int) ([]share.Share, error) {
 	shrLn := int(f.hdr.shareSize)
-	ODSLn := int(f.size()) / 2
+	ODSLn := f.size() / 2
```
Comment (Wondertan) [inline store/file/ods.go:238]: The vars here aren't supposed to be capitalized

**Review by cristaloleg — APPROVED** [review]
